avm/src/vm/opcode.rs

```diff
@@ -3658,8 +3672,21 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
       let ns = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[0])?)?;
       let key = HandlerMemory::fractal_to_string(hand_mem.read_fractal(args[1])?)?;
       let nskey = format!("{}:{}", ns, key);
-      let ds = Arc::clone(&DS);
-      ds.insert(nskey, hm);
+      let ctrl_port = CONTROL_PORT_CHANNEL.get();
+      let ctrl_port = match ctrl_port {
+        Some(ctrl_port) => Some(ctrl_port.borrow().clone()), // TODO: Use thread-local storage
+        None => None,
+      };
+      let is_key_owner = match ctrl_port {
+        Some(ref ctrl_port) => ctrl_port.is_key_owner(&nskey),
+        None => true,
+      };
+      if is_key_owner {
+        let ds = Arc::clone(&DS);
+        ds.insert(nskey, hm);
+      } else {
+        ctrl_port.unwrap().dssetv(&nskey, &hm).await;
```

Comment:
shouldn't we handle the `None` case?

Reply:
So I want to make this effectively infallible, as well -- what these functions are *supposed* to do (and will eventually) is mark a node as down and then try the next node that's supposed to own the data. Worst case it decides the entire cluster is down besides itself and it should just fall back to the `ds.insert(nskey, hm)` logic.

That's not there yet, but I was hoping to make these improvements over time, so we can unblock the work on figuring out how to expose distributed datastore to the anycloud client applications in parallel.

Reply:
gotcha, maybe in the `match` above change it to
```rust
let ctrl_port = match ctrl_port {
  Some(ctrl_port) => ctrl_port.borrow().clone(),
  None => todo!("poll the rest of the cluster, fallback on local"),
};
```
or something like that?

Reply:
Wait, your issue is on the `unwrap` of the `ctrl_port`, not the return value of the `dssetv` call?

That's an impossible scenario. `is_key_owner` can only ever be `false` if `Some(ref ctrl_port)` was the path taken in generating that boolean. Therefore it is 100% guaranteed that `ctrl_port` is a `Some` value and I can `unwrap` it just fine.